### ISSUE TITLE
SE-RTA: runtime assurance, operational envelopes, graceful degradation, and recovery

### ISSUE THREAD

**Issue by Tristan-Stoltz-ERC** [Luminous-Dynamics/symthaea#3779]
## Goal

Generalize Symthaea's strongest existing domain-specific runtime-safety patterns into a reusable **runtime assurance / graceful-degradation contract** for engineered systems, without replacing domain safety cases, ETK, control-system interlocks, or human/institutional authority.

Parent programs: #3677, #3685, #3695, #3697, #3701, #3759.

This program should **extract and compose existing proven patterns** from `symthaea-subterranean` and `symthaea-helicopter` rather than inventing a new universal safety controller.

### Subterranean formal runtime assurance

Current `FORMAL_RUNTIME_ASSURANCE_PROTOCOL.md` already establishes several strong reusable principles:

- a deliberately bounded abstract runtime state;
- transition-contract checking before final-command invariant enforcement;
- same-frame latching of a formal hold;
- **safety-monotonic behavior**: the assurance layer may remove movement/work authority but may not increase actuator command;
- containment/recovery actions remain available while productive authority is held;
- a single clean observation cannot clear a hold; bounded clean dwell at a safe service location is required;
- deterministic replay ledger;
- bounded-model coverage with explicit finite-scope non-claims;
- adversarial mutation tests;
- restart/checkpoint continuity;
- external production qualification remains separate.

### Subterranean post-deployment learning

Current post-deployment protocol already has:

- immutable baseline warmup;
- distribution-shift gating;
- zero-authority shadow candidate;
- held-out non-regression;
- split-role approval;
- bounded canary authority;
- automatic rollback;
- preserved recoverable baseline;
- learning below every established safety authority.

### Helicopter operational/fault monitoring

Current helicopter code already demonstrates:

- configuration-bound operational limit sets;
- `Go / Restricted / NoGo / Incomplete` rather than one boolean;
- missing evidence as a distinct status;
- explicit airframe-configuration mismatch rejection;
- debounced fault detection/isolation;
- real-time monitor sequencing/deadline checks.

These should become reference patterns, not a single cross-domain implementation copied blindly.

### ASTM F3269-21

Use conceptually as a current public reference for Runtime Assurance architectures that safely bound complex/unassured aircraft functions. Do not copy proprietary standard text.

### NASA / FAA Runtime Assurance

Align with the widely used RTA/Simplex structure:

```text
advanced / complex controller
        +
runtime monitor / decision module
        +
trusted reversionary / recovery controller
        +
assured switching / mitigation path
```

NASA also researches multi-monitor RTA and dynamic assurance for autonomous systems.

### AGREE / AADL

AGREE is a compositional assume-guarantee model checker integrated with AADL/OSATE. After SE-009 qualifies, use AGREE as an optional design-time formal source for component contracts that can later project selected executable properties into runtime monitors.

Do not assume:

```text
AGREE proof != runtime-monitor implementation correctness
runtime monitor != AGREE proof
```

## Core laws

```text
monitor triggered != root cause established
monitor quiet != system globally safe
runtime-safe-now != system qualified
fallback engaged != fault diagnosed
recovery succeeded != authority automatically restored
inside operational envelope != requirement satisfaction
outside validated envelope != necessarily immediate physical failure
degraded mode != nominal mode
dynamic assurance update != evidence admission
runtime observation != design-time proof
```

Most importantly:

```text
runtime assurance may reduce authority
!= runtime assurance may mint authority
```

### Operational envelope

```text
OperationalEnvelope {
    envelope_id
    subject_configuration
    model_revision_refs[]
    phase/mode
    prerequisites[]
    required_observation_refs[]
    bounded quantities[]
    validity/evidence_refs[]
    degraded_regions[]
    prohibited_regions[]
    unknown/incomplete policy
}
```

Do not encode one universal safe/unsafe scalar.

### Runtime property

```text
RuntimeProperty {
    property_id
    source_requirement/hazard/control refs
    assumption_refs[]
    observable bindings
    temporal semantics
    evaluation cadence
    reaction_time_budget
    violation semantics
    monitor_revision
}
```

### Monitor result

```text
RuntimeMonitorResult {
    property_ref
    exact observation/state refs
    status
    witness/violation window
    timing diagnostics
    data-quality status
    uncertainty/applicability flags
}
```

A monitor result is an observation/analysis artifact, not an authority capability.

### Degradation / recovery

```text
DegradationMode {
    mode_id
    removed capabilities[]
    retained safety/containment capabilities[]
    entry conditions
    exit conditions
    max dwell?
    operator requirements?
}
```

```text
RecoveryPlan {
    recovery_id
    trigger
    target safe set / service state
    allowed actions
    forbidden actions
    preconditions
    completion evidence
    re-entry policy
}
```

### RTA decision

```text
RtaDecision {
    input monitor results[]
    prior authority state
    decision policy revision
    action:
        ContinueBounded
      | Restrict
      | Hold
      | Isolate
      | SwitchToReversionary
      | Recover
      | Shutdown
      | RequestHumanAuthority
    resulting authority ceiling
    exact reason/witness refs
}
```

The resulting authority ceiling must be <= the pre-existing independently authorized ceiling.

## Authority monotonicity theorem

For the initial generic contract:

```text
RTA output authority <= independently authorized input authority
```

except for a **separate, independently authorized recovery capability** whose permitted actions were already granted before the violation.

Example:

```text
productive motion authority lost
while
cooling / sealing / braking / safe-return authority remains
```

This matches the current subterranean design and avoids the paradox where a monitor alarm grants itself new physical powers.

## Hold / re-entry semantics

Do not clear a violation merely because the latest sample looks good.

Support explicit policies such as:

- latched until human release;
- clean dwell;
- safe-service-location dwell;
- hysteresis;
- independent sensor confirmation;
- fault isolation + repair;
- configuration revision;
- requalification required.

```text
violation absent now != safe to resume nominal authority
```

## Operational envelope and SE-MODEL

SE-MODEL #3759 supplies model applicability / state-estimation context.

Before deciding that a residual is a fault:

```text
check observation quality
-> runtime numerical/monitor health
-> current state-estimation uncertainty
-> applicability/domain shift
-> parameter/model discrepancy
-> fault hypothesis
```

Leaving a validated model envelope may trigger `ApplicabilityReviewRequired`, degraded operation, or reversion without claiming a diagnosed component fault.

## Relationship to SE-OBS

SE-OBS supplies provenance-bound observations.

RTA monitors consume exact observation/state references and emit monitor artifacts. They do not rewrite observations or manufacture measurement truth.

## Relationship to SE-FORMAL

Design-time formal contracts may help establish monitor logic or switching conditions.

Potential later sources:

- TLA+/TLC for mode/authority/recovery state machines;
- Z3/SMT for bounded invariants;
- AGREE/AADL assume-guarantee contracts;
- bounded hybrid/RTA verification where domain-appropriate.

A proved abstraction remains bounded by its abstraction assumptions.

## Relationship to SE-VV

SE-VV owns qualification/credibility questions:

- is the monitor sensitive/specific enough?;
- is reaction time sufficient?;
- does the reversionary controller recover the declared safe set?;
- are common-cause failures between advanced and backup paths controlled?;
- is the operational envelope physically validated?;
- does degraded mode satisfy its own requirements?;
- are re-entry criteria validated?

## Dynamic assurance

Later, project selected runtime facts into a **dynamic assurance view**:

```text
static assurance case
+ exact current configuration
+ current operational envelope
+ monitor/currentness facts
+ active degradation/recovery state
-> dynamic assurance projection
```

This projection must never silently mutate ETK claims. It may request:

- currentness review;
- applicability review;
- evidence refresh;
- operator review;
- re-verification.

## Multi-monitor composition

Do not reduce many monitors to a hidden weighted score.

Need an explicit conflict/composition policy for:

- multiple independent hazards;
- conflicting mitigations;
- monitor dependencies/common sensors;
- stale monitor inputs;
- missing data;
- one monitor requesting continue while another requests hold;
- cascading degradations.

Conservative default:

```text
authority ceiling = meet of applicable independent ceilings
```

with explicit recovery-action exceptions already independently authorized.

### SE-RTA-000 — runtime-assurance architecture profile
Docs only. Freeze authority monotonicity, envelope, monitor, degradation, recovery and re-entry semantics.

### SE-RTA-001 — typed operational-envelope contract
No actuation. Exact configuration/model/evidence bindings and `Nominal/Restricted/Prohibited/Incomplete`-style status.

### SE-RTA-002 — runtime-property / monitor-result contract
Typed observable bindings, temporal semantics, witness traces, stale/missing-data handling.

### SE-RTA-003 — degradation/recovery state machine
Pure state machine only; no hardware interface.

### SE-RTA-004 — authority-ceiling algebra
Prove the generic composition cannot increase authority. Include multi-monitor composition and retained recovery capabilities.

### SE-RTA-005 — subterranean reference adapter
Project existing formal-assurance/hold/recovery behavior into the generic types without weakening existing domain code.

### SE-RTA-006 — helicopter reference adapter
Project operational-limit, FDI, and real-time monitor results into the generic contract.

### SE-RTA-007 — replay / restart / continuity receipts
Reuse existing deterministic replay and checkpoint principles; bind monitor state and authority state across restart.

### SE-RTA-008 — formal monitor sources
Optional TLA+/SMT/AGREE projections after SE-FORMAL/SE-009 qualify.

### SE-RTA-009 — HIL fault/recovery corpus
Sensor/actuator/timing/power/communications/configuration faults; reaction time; common-cause and backup-path failures.

### SE-RTA-010 — dynamic assurance projection
Read-only linkage into assurance/currentness review; no ETK mutation.

### SE-RTA-011 — post-deployment adaptation composition
Compose #3759 shadow-model/controller promotion with RTA: candidate learning remains below independent envelopes/interlocks and preserves immediate baseline rollback.

## Qualification corpus

At minimum include:

- one-frame transient must not spuriously reconfigure where debounce policy requires persistence;
- persistent violation must trigger within bounded reaction time;
- stale/missing input;
- contradictory monitors;
- monitor itself late/unhealthy;
- advanced controller fault;
- reversionary-controller fault;
- shared-sensor/common-cause failure;
- switch/reversion failure;
- recovery action unavailable;
- clean sample after violation must not improperly clear latch;
- restart during hold;
- restart during recovery;
- configuration mismatch;
- operational envelope becomes stale after model/configuration change;
- candidate learned controller attempts authority outside canary;
- all nominal guidance unavailable but independent containment action remains available.

## Non-goals

- no universal controller;
- no universal safety envelope;
- no claim that a monitor proves physical safety;
- no assumption that fallback is fault-free;
- no runtime-monitor -> ETK evidence shortcut;
- no autonomous authority creation;
- no replacement for domain certification;
- no reuse of consciousness/FEP surprise thresholds as physical safety limits;
- no single risk/readiness/confidence scalar.

## Exit criterion

Symthaea can represent and verify the runtime **contract around** a complex controller: exact operating envelope, monitor properties, degraded modes, trusted fallback/recovery paths, authority reductions, replay/restart continuity, and re-entry criteria—while preserving the fact that runtime assurance bounds behavior but does not itself certify the plant, diagnose every fault, or grant new authority.

**Comment by Tristan-Stoltz-ERC** [comment]
Deeper refinement: avoid overloading one "operational envelope".

## Four distinct envelopes

### 1. Physical operating envelope

Externally supported limits for the actual configured system:

```text
temperature
load / mass / CG
pressure
speed
wind / weather
power / thermal
structural / environmental limits
```

This is closest to the existing helicopter `OperationalLimitSet`.

Crossing it may require restriction/no-go/recovery, but the response remains domain-specific.

### 2. Model applicability envelope

From SE-MODEL / SE-VV:

```text
where this exact model revision has established credibility
```

Leaving the model-applicability envelope means predictions/estimates may no longer be supportable. It does **not** necessarily mean the physical system has already crossed a physical limit.

```text
out of model envelope != physically unsafe
```

but it may require degraded authority because decision support has weakened.

### 3. Recoverable / invariant envelope

Control/RTA concept:

```text
states from which the trusted recovery path can still
return/hold the system inside a declared safe set
within reaction/resource constraints
```

This can be narrower than the physical envelope.

The monitor may need to switch/restrict **before** a hard physical limit is violated.

Future types might distinguish:

```text
SafeSet
RecoverableSet
PointOfNoGuaranteedReturn
MinimumRiskCondition
```

Do not imply these are analytically known for every system.

### 4. Authority envelope

The actions currently allowed by independently issued capabilities / operator / mission / institutional constraints.

```text
physically possible
!= model-supported
!= recoverable
!= authorized
```

RTA must never widen this envelope.

## Recommended composition

A runtime decision should bind all four independently:

```text
PhysicalEnvelopeStatus
ModelApplicabilityStatus
RecoverabilityStatus
AuthorityEnvelope
        ↓
RtaDecision
```

Do not collapse them into one "safe score".

## Monitor assurance / common cause

The monitor itself is a safety-critical component and must expose health:

- observation freshness / clock quality;
- computation deadline;
- internal state continuity;
- configuration/property revision;
- sensor/model dependencies;
- self-test status where available;
- independence/shared-root metadata;
- degraded-monitor policy.

Required cases:

```text
controller healthy + monitor unhealthy
controller unhealthy + monitor healthy
controller and monitor share failed sensor
advanced + backup controllers share software/model defect
switch actuator fails
```

A quiet or failed monitor must never be interpreted as evidence that no hazard exists.

## Reversion is not enough

For Simplex-like paths, verify separately:

1. detection before loss of recoverability;
2. switch latency;
3. reversionary controller availability;
4. recovery/containment effectiveness;
5. retained energy/resources for recovery;
6. no harmful transient caused by switching;
7. re-entry is independently gated.

This suggests a future `RecoverabilityAssessment` / `ReversionReceipt` rather than one `fallback=true` flag.

## Method posture

Control-barrier functions, reachability/viability analysis, hybrid verification, MPC safety filters, and other methods may help establish safe/recoverable sets in appropriate domains. Treat these as **analysis providers**, not universal RTA semantics.

```text
computed safe set != physically validated safe set
```

The generic SE-RTA layer should store the exact method/model/evidence identity and remain agnostic to how the set was obtained.

**Comment by Tristan-Stoltz-ERC** [comment]
Add explicit RTA evaluation metrics; do not use a single resilience/safety score.

## Detection / monitor quality

Measure separately where ground truth or controlled injection exists:

- time-to-detect;
- missed-detection rate;
- false-trip rate;
- stale-input detection;
- monitor-deadline miss rate;
- property/witness localization quality;
- debounce/hysteresis behavior.

A monitor that trips constantly is not "safer" merely because it catches every injected fault.

## Mitigation / recovery quality

- detection-to-authority-reduction latency;
- switch/reversion latency;
- switch success rate;
- peak excursion after trigger;
- safe-set / validated-bound margin where defined;
- time-to-recovery / safe-service state;
- recovery resource consumption;
- retained minimum capabilities;
- failed recovery rate;
- re-entry false-clear rate.

## Availability / mission impact

Keep distinct from safety:

- nominal availability;
- restricted/degraded dwell;
- nuisance shutdowns;
- mission loss / productivity reduction;
- operator intervention load.

This prevents a trivially "safe" controller that always shuts down from winning hidden scalar optimization.

## Independence / common-cause

Report:

- shared sensor roots;
- shared estimator/model roots;
- shared software/library roots;
- shared power/compute/network roots;
- shared training data where learned components are involved;
- monitor/advanced/reversionary controller diversity profile.

Do not convert "different process" or "different model name" into independence.

## Paired RTA effectiveness experiments

For preregistered synthetic/HIL fault cases, where safe to do so, compare:

```text
same frozen initial state + same injected fault
    RTA active
vs
    RTA bypassed in simulation/test harness only
```

Measure whether RTA changes:
- hazard excursion;
- time outside limits;
- recoverability;
- final safe state;
- mission impact.

The bypass arm is an **experimental control only** and must never become a deployable runtime mode for safety-critical operation.

## Strong claim boundary

```text
RTA reduced harm in tested faults
!= all faults covered
!= fallback universally safe
!= system certified
```

Also retain negative results: an RTA architecture that worsens an excursion, switches too late, or creates harmful transients is a valid and important finding.
